src/emulator/apphosting/serve.ts

```diff
@@ -2,37 +2,55 @@
  * Start the App Hosting server.
  * @param options the Firebase CLI options.
  */
-import { isIPv4 } from "net";
-import { checkListenable } from "../portUtils";
-import { wrapSpawn } from "../../init/spawn";
+
+import { spawn } from "cross-spawn";
+import { discoverPackageManager } from "./utils";
 
 /**
  * Spins up a project locally by running the project's dev command.
  */
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export async function start(options: any): Promise<{ port: number }> {
-  let port = options.port;
-  while (!(await availablePort(options.host, port))) {
-    port += 1;
-  }
 
-  serve(options, port);
+export async function start(): Promise<{ hostname: string; port: number }> {
+  const hostUrl = await serve();
 
-  return { port };
+  const { hostname, port } = new URL(hostUrl);
+  return { hostname, port: +port };
 }
 
-function availablePort(host: string, port: number): Promise<boolean> {
-  return checkListenable({
-    address: host,
-    port,
-    family: isIPv4(host) ? "IPv4" : "IPv6",
+async function serve(): Promise<string> {
+  const rootDir = process.cwd();
+  const packageManager = await discoverPackageManager(rootDir);
+
+  /**
+   *  TODO: Should add a timeout here. If the hostUrl cannot be retrieved
+   *  in a timely manner, user is probably using an unsupported framework.
+   * */
+
+  const hostUrl = new Promise<string>((resolve, reject) => {
+    const serve = spawn(packageManager, ["run", "dev"], {
+      cwd: rootDir,
+    });
+
+    serve.stdout.on("data", (data: any) => {
+      process.stdout.write(data);
+      const url = getHostUrlFromString(data.toString());
+      if (url) resolve(url);
+    });
+    serve.stderr.on("data", (data: any) => {
+      process.stderr.write(data);
+    });
+    serve.on("exit", reject);
   });
+
+  return await hostUrl;
 }
 
 /**
  * Exported for unit testing
  */
-export async function serve(options: any, port: string) {
-  // TODO: update to support other package managers and frameworks other than NextJS
-  await wrapSpawn("npm", ["run", "dev", "--", "-H", options.host, "-p", port], process.cwd());
+export function getHostUrlFromString(data: string): string | undefined {
```

Comment:
Can you clarify what this function is doing? It looks like it's doing few things like

1) ignoring URLs that starts with https://  (why?)
2) stripping out "/" (why?)

Reply:
outdated - not using the approach of reading host/port from STDOUT anymore